### PR TITLE
Non-autoregressive ASR with Mask CTC

### PR DESCRIPTION
**Pull request by YosukeHiguchi**
Implementation of https://arxiv.org/abs/2005.08700

### CODE REVIEW

File: espnet/bin/asr_recog.py [hunk 1]
```diff
@@ -206,6 +206,19 @@ def get_parser():
...
+        "--p-thres",
+        type=float,
+        default=0.999,
+        help="Threshold probability for CTC output"
+    )
+    parser.add_argument(
+        "--K",
+        type=int,
+        default=1,
+        help="Number of decoding iterations. Set to 999 to predict 1 mask/iter."
```
Comment (sw005320) [inline espnet/bin/asr_recog.py:220]: set 0 instead of 999?

File: espnet/asr/pytorch_backend/asr.py [hunk 2]
```diff
@@ -408,6 +408,8 @@ def train(args):
         int(valid_json[utts[0]]["input"][i]["shape"][-1]) for i in range(args.num_encs)
     ]
     odim = int(valid_json[utts[0]]["output"][0]["shape"][-1])
+    if args.dmode == "NAR":
+        odim += 1
```
Comment (sw005320) [inline espnet/asr/pytorch_backend/asr.py:412]: please add a comment of why you increase the odim

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py [hunk 3]
```diff
@@ -199,6 +201,15 @@ def add_arguments(parser):
         group.add_argument(
             "--dunits", default=320, type=int, help="Number of decoder hidden units"
         )
+        # Non-autoregressive training
+        group.add_argument(
+            "--dmode",
```
Comment (sw005320) [inline espnet/nets/pytorch_backend/e2e_asr_transformer.py:206]: what does "dmode" mean?

File: espnet/bin/asr_recog.py (hunk 1 above)
Comment (sw005320) [reply espnet/bin/asr_recog.py:220]: Maybe you can set an appropriate default value than 1?

File: espnet/bin/asr_train.py [hunk 4]
```diff
@@ -596,6 +596,8 @@ def main(cmd_args):
         char_list = [entry.decode("utf-8").split(" ")[0] for entry in dictionary]
         char_list.insert(0, "<blank>")
         char_list.append("<eos>")
+        if args.dmode == "NAR":
+            char_list.append("<mask>")
```
Comment (sw005320) [inline espnet/bin/asr_train.py:600]: please add an explanation about `<mask>`

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py (hunk 3 above)
Comment (YosukeHiguchi) [reply espnet/nets/pytorch_backend/e2e_asr_transformer.py:206]: It means "decoder training mode", specifying whether decoder is trained in the autoregressive or non-autoregressive manner.

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py [hunk 5]
```diff
@@ -317,11 +334,18 @@ def forward(self, xs_pad, ilens, ys_pad):
...
+                    ys_pad, self.mask, self.eos, self.ignore_id
+                )
+                ys_mask = (ys_in_pad != self.ignore_id).unsqueeze(-2)
+                pred_pad, pred_mask = self.decoder(ys_in_pad, ys_mask, hs_pad, hs_mask)
+            else:
+                ys_in_pad, ys_out_pad = add_sos_eos(
+                    ys_pad, self.sos, self.eos, self.ignore_id
+                )
+                ys_mask = target_mask(ys_in_pad, self.ignore_id)
+                pred_pad, pred_mask = self.decoder(ys_in_pad, ys_mask, hs_pad, hs_mask)
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/e2e_asr_transformer.py:348]: This is the same as `line:342`.

File: espnet/asr/pytorch_backend/asr.py (hunk 2 above)
Comment (hirofumi0810) [inline espnet/asr/pytorch_backend/asr.py:412]:
```suggestion
        odim += 1  # for the <mask> token
```

File: espnet/asr/pytorch_backend/asr.py [hunk 6]
```diff
@@ -976,6 +978,10 @@ def recog(args):
                             for n in range(args.nbest):
                                 nbest_hyps[n]["yseq"].extend(hyps[n]["yseq"])
                                 nbest_hyps[n]["score"] += hyps[n]["score"]
+                elif model.dmode == 'NAR':
+                    nbest_hyps = model.recognize_mask_ctc(
```
Comment (hirofumi0810) [inline espnet/asr/pytorch_backend/asr.py:982]:
```suggestion
                    nbest_hyps = model.recognize_nat(
```
For future research with other NAT variants, it might be better to generalize the name.

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py [hunk 7]
```diff
@@ -613,6 +637,67 @@ def recognize(self, x, recog_args, char_list=None, rnnlm=None, use_jit=False):
         )
         return nbest_hyps
 
+    def recognize_mask_ctc(self, x, recog_args, char_list=None):
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/e2e_asr_transformer.py:640]:
```suggestion
    def recognize_nat(self, x, recog_args, char_list=None):
```
And you can call `recognize_mask_ctc` inside `recognize_nat`.

File: espnet/bin/asr_train.py (hunk 4 above)
Comment (hirofumi0810) [reply espnet/bin/asr_train.py:600]: This is probably problematic because RNN models do not have `dmode` in args.

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py [hunk 8]
```diff
@@ -254,8 +265,14 @@ def __init__(self, idim, odim, args, ignore_id=-1):
         else:
             self.decoder = None
         self.blank = 0
-        self.sos = odim - 1
-        self.eos = odim - 1
+        self.dmode = args.dmode 
+        if self.dmode == "NAR":                                                                                                                                                                
+            self.mask = odim - 1
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/e2e_asr_transformer.py:270]: `self.mask` is misleading. Please consider `self.mask_token`

File: espnet/nets/pytorch_backend/transformer/add_sos_eos.py [hunk 9]
```diff
@@ -29,3 +30,35 @@ def add_sos_eos(ys_pad, sos, eos, ignore_id):
     ys_in = [torch.cat([_sos, y], dim=0) for y in ys]
     ys_out = [torch.cat([y, _eos], dim=0) for y in ys]
     return pad_list(ys_in, eos), pad_list(ys_out, ignore_id)
+
+def mask_uniform(ys_pad, mask, eos, ignore_id):
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/transformer/add_sos_eos.py:34]:
```suggestion
def mask_uniform(ys_pad, mask_token, eos, ignore_id):
    """Replace random tokens with <mask> label and add <eos> label.

    The number of <mask> is chosen from a uniform distribution
    between one and the target sequence's length.
    :param torch.Tensor ys_pad: batch of padded target sequences (B, Lmax)
    :param int mask_token: index of <mask>
    :param int eos: index of <eos>
    :param int ignore_id: index of padding
    :return: padded tensor (B, Lmax)
    :rtype: torch.Tensor
    :return: padded tensor (B, Lmax)
    :rtype: torch.Tensor
    """
    from espnet.nets.pytorch_backend.nets_utils import pad_list

    ys = [y[y != ignore_id] for y in ys_pad]  # parse padded ys
    ys_out = [y.new(y.size()).fill_(ignore_id) for y in ys]
    ys_in = [y.clone() for y in ys]
    for i in range(len(ys)):
        num_samples = numpy.random.randint(1, len(ys[i]) + 1)
        idx = numpy.random.choice(len(ys[i]), num_samples)

        ys_in[i][idx] = mask_token
        ys_out[i][idx] = ys[i][idx]

    return pad_list(ys_in, eos), pad_list(ys_out, ignore_id)
```

File: espnet/nets/pytorch_backend/transformer/add_sos_eos.py [hunk 10]
```diff
@@ -29,3 +30,35 @@ def add_sos_eos(ys_pad, sos, eos, ignore_id):
...
+    ys_out = [y.clone() * 0 + ignore_id for y in ys]
+    ys_in = [y.clone() for y in ys]
+    for i in range(len(ys)):
+        num_samples = numpy.random.randint(1, len(ys[i]) + 1)
+        idx = numpy.random.choice(len(ys[i]), num_samples)
+
+        ys_in[i][idx] = mask
+        ys_out[i][idx] = ys[i][idx]
+
+    return pad_list(ys_in, eos), pad_list(ys_out, ignore_id)
```
Comment (hirofumi0810) [inline espnet/nets/pytorch_backend/transformer/add_sos_eos.py:64]:
```suggestion
def mask_uniform(ys_pad, mask_token, eos, ignore_id):
    """Replace random tokens with <mask> label and add <eos> label.
    The number of <mask> is chosen from a uniform distribution
    between one and the target sequence's length.
    :param torch.Tensor ys_pad: batch of padded target sequences (B, Lmax)
    :param int mask_token: index of <mask>
    :param int eos: index of <eos>
    :param int ignore_id: index of padding
    :return: padded tensor (B, Lmax)
    :rtype: torch.Tensor
    :return: padded tensor (B, Lmax)
    :rtype: torch.Tensor
    """
    from espnet.nets.pytorch_backend.nets_utils import pad_list
    ys = [y[y != ignore_id] for y in ys_pad]  # parse padded ys
    ys_out = [y.new(y.size()).fill_(ignore_id) for y in ys]
    ys_in = [y.clone() for y in ys]
    for i in range(len(ys)):
        num_samples = numpy.random.randint(1, len(ys[i]) + 1)
        idx = numpy.random.choice(len(ys[i]), num_samples)
        ys_in[i][idx] = mask_token
        ys_out[i][idx] = ys[i][idx]
    return pad_list(ys_in, eos), pad_list(ys_out, ignore_id)
```

File: espnet/asr/pytorch_backend/asr.py [hunk 11]
```diff
@@ -408,6 +408,8 @@ def train(args):
         int(valid_json[utts[0]]["input"][i]["shape"][-1]) for i in range(args.num_encs)
     ]
     odim = int(valid_json[utts[0]]["output"][0]["shape"][-1])
+    if hasattr(args, 'dmode') and args.dmode == "NAR":
```
Comment (sw005320) [inline espnet/asr/pytorch_backend/asr.py:411]: Can you change `dmode` to `decoder-mode`?

File: espnet/bin/asr_recog.py [hunk 12]
```diff
@@ -206,6 +206,19 @@ def get_parser():
...
     )
+    # non-autoregressive related
+    parser.add_argument(
+        "--p-thres",
+        type=float,
+        default=0.999,
+        help="Threshold probability for CTC output"
+    )
+    parser.add_argument(
+        "--K",
```
[the remark this answers is not in the document]
Comment (sw005320) [reply espnet/bin/asr_recog.py]: sorry. `nar` --> `mp`?

File: espnet/bin/asr_recog.py [hunk 13]
```diff
@@ -206,6 +206,19 @@ def get_parser():
     parser.add_argument(
         "--streaming-offset-margin", type=int, default=1, help="Offset margin"
     )
+    # non-autoregressive related
+    parser.add_argument(
+        "--p-thres",
```
[the remark this answers is not in the document]
Comment (sw005320) [reply espnet/bin/asr_recog.py]: nar --> mp

**Comment by sw005320** [comment]
Maybe `nar` -> `mp` for most of the parts.

File: egs/wsj/asr1/conf/tuning/decode_pytorch_transformer_mp.yaml [hunk 14]
```diff
@@ -0,0 +1,2 @@
+nar-n-iterations: 10
```
Comment (sw005320) [inline egs/wsj/asr1/conf/tuning/decode_pytorch_transformer_mp.yaml:1]:
```suggestion
mp-n-iterations: 10
```

File: espnet/bin/asr_recog.py [hunk 15]
```diff
@@ -206,6 +206,22 @@ def get_parser():
     parser.add_argument(
         "--streaming-offset-margin", type=int, default=1, help="Offset margin"
     )
+    # non-autoregressive related
+    parser.add_argument(
+        "--nar-n-iterations",
```
Comment (sw005320) [inline espnet/bin/asr_recog.py:211]:
```suggestion
        "--mp-n-iterations",
```

File: espnet/nets/pytorch_backend/e2e_asr_transformer.py [hunk 16]
```diff
@@ -203,6 +205,15 @@ def add_arguments(parser):
...
             "--dunits", default=320, type=int, help="Number of decoder hidden units"
         )
+        # Non-autoregressive training
+        group.add_argument(
+            "--decoder-mode",
+            default="AR",
+            type=str,
+            choices=["AR", "MP"],
+            help="AR: standard autoregressive training, "
+            "MP: non-autoregressive training based on mask-predict",
```
Comment (sw005320) [inline espnet/nets/pytorch_backend/e2e_asr_transformer.py:215]:
```suggestion
            "MP: non-autoregressive training based on mask-predict (MP)",
```

File: egs/wsj/asr1/conf/tuning/decode_pytorch_transformer_mp.yaml (hunk 14 above)
Comment (YosukeHiguchi) [reply egs/wsj/asr1/conf/tuning/decode_pytorch_transformer_mp.yaml:1]: @sw005320 I was thinking of using `nar-n-iterations` for other nar models. Should I still make it specific to the mask-predict model?

Comment (sw005320) [reply egs/wsj/asr1/conf/tuning/decode_pytorch_transformer_mp.yaml:1]: let's make it mask-predict specific for now.
I'm thinking of the more modularized implementation of this one in espnet2 in the near future.

**Comment by sw005320** [comment]
Can you apply `black` to fix the CI error in https://travis-ci.org/github/espnet/espnet/builds/711066906#L1929

**Comment by sw005320** [comment]
Many thanks, @YosukeHiguchi !!!
This is great PR.

Let's further develop this technique.
I found a lot of potentials in this technique!

- [ ] refactoring. I want to make this as a module rather than one inside the transformer module
- [ ] documentation (featuring it in the main `README.md` and add usages)

There is some unreasonable CI errors. I'm not sure why it happens but @YosukeHiguchi, could you merge the latest to make sure?

**Comment by YosukeHiguchi** [comment]
@sw005320 
Thank you for your support!
Yes this PR is the latest.
